### PR TITLE
standard library: implement `help` commands with `$nu`

### PR DESCRIPTION
**Pull request by amtoine**
Related to one task in #8311.

# Description
this PR implements the built-in `help` commands in the standard library using the `$nu` variable
as the only source to access information about the commands, aliases and modules.
this can be thought as a proof of concept and a replacement proposition in the long-run, as the
`help` commands are not performance-critical and might not require taking up compilation time
and space.

> **Note**
> this PR has in mind the goal of cloning the builtin `help` commands.
> this is what i've tried as much as possible with the information available in `$nu` :relieved:
>
> **However**, some things appear impossible right now, see the **Warning** below.

## progress
- [x] `help modules`
- [x] `help aliases`
- [x] `help externs`  :warning: might require augmenting `$nu` with external flags
- [x] `help operators`  :heavy_check_mark: ~~might require augmenting `$nu` with the table of operators~~ => done in 55350c124e0f4317bc4aab12259878a010d4fdc4
- [x] `help commands`
- [x] `help` :heavy_check_mark: ~~can not find a way to implement that...~~ => done in [`a387256b4453cd8c04c28dce002e301da7d06157`..`62cf425efd8d5907115b5fc7f17ecccfe3c2763e`](a387256b4453cd8c04c28dce002e301da7d06157..62cf425efd8d5907115b5fc7f17ecccfe3c2763e)

### the error system
the `*-not-found-error` are here as convenience wrappers to throw all required errors, i.e. when
a user parameter does not correspond to any commands, alias or module.

### the actual `help` commands
the implemented `help` commands follow the same structure
- use `print-help-header` to print the section names in the `help` messages
- define a `show-*` command that takes the command, alias or module information in the form of a
record and prints as much as possible in the same format as the built-ins
- define a `help *` each having the same signature and behaviour
  - get the whole collections of commands, aliases or modules, sorted by name
  - if `--find` is used, matches the name against the pattern
    - if the result is alone, expand it
    - if not a single result, show the raw table
  - if `--find` is not used but a name is still provided, show it or throw a `*-not-found-error`
  - otherwise show all the commands, aliases or modules

# User-Facing Changes
users can now access the `help` commands via `use std.nu` and not rely on the builtins anymore.

# Tests + Formatting
i do not really know what to write here...
should i take a few know commands and compare the outputs of the `help` on them?

# After Submitting
i'd say probably part of the standard library announcement :thumbsup:

### CODE REVIEW

**Comment by amtoine** [comment]
there it is finally :tada: 

i'd love some feedback on that :yum: 

### my main concerns
- the two :red_circle: points above
- the :orange_circle: point above
- the flags of externals that are not in `$nu`
- the `help` command => i have a draft for this, but it does not work that well for now: **should i keep working on this?**

**Comment by jntrnr** [comment]
@amtoine for 8) above, I think my version of `print_help` can be improved. Instead of using the string interpolation like I did, you can use:

```
for line in ($example.result | table | lines) { print $"  ($line)" }
```

**Comment by amtoine** [comment]
> @amtoine for 8) above, I think my version of `print_help` can be improved.
>
oh yeah i started tweak your script a bit, quite a solid base for this task of printing the `help` pages :relieved: 

> Instead of using the string interpolation like I did, you can use:
> 
> ```
> for line in ($example.result | table | lines) { print $"  ($line)" }
> ```
>
very very nice :ok_hand: 

issue 8 solved in 9b8f07f7d0c4a156fc2d8e1504b8db8b76a0b411

**Comment by jntrnr** [comment]
Is this close enough to land? Or would you like to keep improving it?

**Comment by amtoine** [comment]
> Is this close enough to land? Or would you like to keep improving it?
>
i'd say it's mostly ready apart from
- the few issues in the warning above
- `help externs` which misses a lot of features
- the main `help`

i mostly wanted feedback on these, but we can land this and address these later.
e.g. i can move all these points in an issue :+1: 

as you prefer, i'm pretty with them, except `help externs` and `help` :relieved:

**Comment by jntrnr** [comment]
@amtoine - for the main `help`, what's the main part causing problems?

**Comment by jntrnr** [comment]
(for what it's worth - I'm okay with -f not being exactly the same as it is now)

**Comment by amtoine** [comment]
> @amtoine - for the main `help`, what's the main part causing problems?
>
mm i'm not sure i remember :thinking: 
let me try again real quick :+1:

**Comment by amtoine** [comment]
okey i've tried something here :+1: 

it's not a 100% clone of the builtin `help` for now but it's not super far.
the main things
- speed: it's quite slower than the builtin => not necessarily an issue for `help` commands
- subcommands: with the way the signature expand in `show-command`, `help help` does not list the subcommand of `help` in its help page, see
```bash
use std.nu [
    help
    "help aliases"
    "help commands"
    "help modules"
    "help externs"
]

help help
```

**Comment by amtoine** [comment]
i think we're close :open_mouth: 

can't see much differences between the builtins and the stdlib ones now :thinking:

**Comment by amtoine** [comment]
i moved the "warnings" out of the main message as they are mostly addressed now :+1: 

> **Warning**
> some disclaimers, ideas, caveats and limitations about the implementations above
>
> > legend
> > :white_circle: not a real issue
> > :large_blue_circle: an idea
> > :green_circle: done in some commit
> > :yellow_circle: -> :orange_circle: -> :red_circle: from low to high issue
>
> 1. :white_circle: `--find` only tries to match in `$nu.scope.*.name` => the built-in match on the name and the usage
> 2. :large_blue_circle: i currently print the exported environment blocks, see `help help modules`
> 3. :large_blue_circle: i thought of adding a `--raw` option to allow custom data processing by not expanding the help page found, but instead return the raw data
> 4. :green_circle: multi word ~~commands, aliases and modules does not match => they have to be put in quotes~~ => done in 9fce454da534bb7195b1d868b5b47928b31c6b77
> 5. :white_circle: ~~`help aliases ...` does not say "This alias wraps an unknown external command."~~ => but `help <some alias>` does
> 6. :yellow_circle: with the builtin, `help externs open` returns the `help open` even though `open` is not an external => the new `std help externs open` does not
> 7. :red_circle: in the built-in `help externs`, we have all the registered flags that show up => as they are not available in `$nu`, i do not know how to do that 
> 8. :green_circle: in `help commands`, we do not get pretty example results... => done in 9b8f07f7d0c4a156fc2d8e1504b8db8b76a0b411 
> e.g. for the `merge` command, we get
> ```bash
> Examples:
>   Add an 'index' column to the input table
>   > [a b c] | wrap name | merge ( [1 2 3] | wrap index )
>   [{name: a, index: 1}, {name: b, index: 2}, {name: c, index: 3}]
> 
>   Merge two records
>   > {a: 1, b: 2} | merge {c: 3}
>   {a: 1, b: 2, c: 3}
> 
>   Merge two tables, overwriting overlapping columns
>   > [{columnA: A0 columnB: B0}] | merge [{columnA: 'A0*'}]
>   [{columnA: A0*, columnB: B0}]
> ```
> instead of
> ```bash
> Examples:
>   Add an 'index' column to the input table
>   > [a b c] | wrap name | merge ( [1 2 3] | wrap index )
>   ╭───┬──────╮
>   │ # │ name │
>   ├───┼──────┤
>   │ 1 │ a    │
>   │ 2 │ b    │
>   │ 3 │ c    │
>   ╰───┴──────╯
> 
>   Merge two records
>   > {a: 1, b: 2} | merge {c: 3}
>   ╭───┬───╮
>   │ a │ 1 │
>   │ b │ 2 │
>   │ c │ 3 │
>   ╰───┴───╯
> 
>   Merge two tables, overwriting overlapping columns
>   > [{columnA: A0 columnB: B0}] | merge [{columnA: 'A0*'}]
>   ╭───┬─────────┬─────────╮
>   │ # │ columnA │ columnB │
>   ├───┼─────────┼─────────┤
>   │ 0 │ A0*     │ B0      │
>   ╰───┴─────────┴─────────╯
> ```
> 9. :green_circle: `help commands` does not tell when a parameter is optional => done in d0a7a14d6852afdc787699d976d1a1e3b42958b2
> 10. :green_circle: `help commands` does not list the true signatures => done in 656bfa4462785793a72bc34ae80fbdb39776ab25
> e.g. the `open` command should show
> ```bash
> Signatures:
>   <nothing> | open <string?> -> <any>
>   <string> | open <string?> -> <any>
> ```
> 11. :green_circle: `help commands` does not list subcommands => done in ad7f4c417af6c0c512df66fc66b1253c41c6fb5a

**Comment by amtoine** [comment]
## final word
i think we're good :+1: 

there will be more work and effort to improve potential bugs and missing features, but it looks quite good to me right now :yum: 

i'd say i've got two things in mind
- the tests
- add information about the externals in `$nu` to complete `help externs`

i think i'll open an issue about completing these once this is merged :relieved:

**Comment by fdncred** [comment]
This is an impressive amount of work. Good job! This will be a good experiment.

Is this still the case? If so I think we need to work on it a bit more.
```
--find only tries to match in $nu.scope.*.name => the built-in match on the name and the usage
```
The whole purpose of adding search terms was being able to search them. The current `help -f` searches `name`, `usage`, and `search_terms`. I think we need to maintain that.

Another idea is that I'd like to run both of these side by side. I wonder if we can initially make these something like `std help blah` so we have the stdlib help commands and the regular help commands. I want to do some benchmarks and compare/contrast them as well.

**Comment by fdncred** [comment]
> i think there is a search_terms column in $nu.scope.commands only, and as you can see [here](https://github.com/amtoine/nushell/blob/feature/implement-help-commands-in-standard-library/crates/nu-utils/standard_library/std.nu#L932)

Right but we need name, usage, and search_terms searched. Are you saying all 3 of those are searched?

> the help commands are not removed from the rust source, maybe the easiest, to avoid giving them dumb names merely for benchmarking, might just be to open two shells where one overwrites the help commands with

But we won't be able to do that if the prelude includes the help commands. this is why i'm saying that the std.nu commands should be named something different or `use`d differently so you can do `std help something` and `help something`. Once we all agree on liking the scripts, we can change how they're used and no one will ever know the difference.

**Comment by fdncred** [comment]
I just noticed that `help -f arg` doesn't highlight the hits. I was expecting it to look like this
image

**Comment by fdncred** [comment]
Any ideas why `help externs` seem to be double spaced?
image

**Comment by fdncred** [comment]
`help commands` is returning too many columns compared to the built-in

**Comment by fdncred** [comment]
`help` output really looks nice. good job.
`help help` is a bit overwhelming compared to the built-in

**Comment by amtoine** [comment]
> Right but we need name, usage, and search_terms searched. Are you saying all 3 of those are searched?
> 
with 2e6324188bbdc4a1e65560c49761cd2972883534 yes :wink: 

> But we won't be able to do that if the prelude includes the help commands. this is why i'm saying that the std.nu commands should be named something different or `use`d differently so you can do `std help something` and `help something`. Once we all agree on liking the scripts, we can change how they're used and no one will ever know the difference.
>
agree :+1: 
i'm not a huge fan of renaming them here BUT here is what we can do:
- once the prelude fully works in #8627, we will be able to rename the commands such that they are accessed as `std help ...` without modifying their names in the standard library!
- we can wait more to land #8611, once we decide if the transition can be made, and until that point, you'll have access to the builtins as `help ...` and to those of the standard library as `std help ...` once the prelude from #8627 is ready (very soon? :innocent:)

would that be ok with what you want to test? :yum:

**Comment by amtoine** [comment]
> I just noticed that `help -f arg` doesn't highlight the hits.
>
mm yeah i know...

this is because
- `find` shows the hits in red => but we cannot find in specific columns (can we?)
- `where` allows us to only search in `name`, and `usage` and `search_terms` when available => it does not highlight the hits

there might be a tradeoff to make here? :thinking: 


> **Note**
> reproduce this with the `cargo` example you gave above 
> - `$nu.scope.commands | where is_extern | find cargo` will give a ton of of `red` BUT will search in all columns
> - `$nu.scope.commands | where is_extern | where name =~ cargo or usage =~ cargo or search_terms =~ cargo` does not give colors but only gives result matching the desired columns

**Comment by amtoine** [comment]
> Any ideas why `help externs` seem to be double spaced?
>
yeah not sure why externs do that...
fixed in 0f809c95778b78124ef9b6d42d88001744958c36 :+1:

**Comment by amtoine** [comment]
> `help` output really looks nice. good job.
>
:blush: 
not enough colors in the builtin :sparkles:

> `help help` is a bit overwhelming compared to the built-in
>
that's because it follows the exact same template as the other commands, i.e. it's built from `$nu.scope.commands` :thinking: 

do you want to write a special `help help` command to make its output less overwhelming as it might be on of the first commands a new user runs?

**Comment by amtoine** [comment]
@fdncred i think i've answered to everything :relieved:

**Comment by amtoine** [comment]
> Related to `help -f arg` and coloring. We could do something like this
> 
> ```
> $nu.scope.commands | find arg | reject module_name examples is_builtin is_sub is_plugin is_custom is_keyword is_extern creates_scope extra_usage
> ```
> 
with this, we gain the highlight but, as i said, we not only search the name, usage and search terms :thinking: 
do we want
- hightlighting with `find`
- narrower search to avoid hitting garbase matches
?

> but i think the signatures would need to be updated. Something weird is going on with signatures here recently.
> 
> Note the signature in the later are like `<nothing> | bytes build --> <binary>`
>
oh yeah you're right :thinking: 
that's `$nu` that needs an update for this, right?

**Comment by fdncred** [comment]
> help -f

I definitely want highlighting and maybe a flag to turn it off. If we don't do it with find we'll need to do it manually by finding the index-of and str replace. It will probably be slow, but I'd prefer to have identical output of what we have now.

> that's $nu that needs an update for this, right?

I'm not sure that's stored anywhere right now just like this. You'd have to `get signatures` then `get syntax_shape | where parameter_type == input` and something similar for output in order to construct something like exists now.

If we're trying to exactly match `help commands` we'll also have to add `command_type`.

**Comment by amtoine** [comment]
> I definitely want highlighting and maybe a flag to turn it off. If we don't do it with find we'll need to do it manually by finding the index-of and str replace. It will probably be slow, but I'd prefer to have identical output of what we have now.
> 
okey, i add that back :+1: 
 
> I'm not sure that's stored anywhere right now just like this. You'd have to `get signatures` then `get syntax_shape | where parameter_type == input` and something similar for output in order to construct something like exists now.
> 
mm we do not have this kind of output with the new `help` commands.
`$nu.scope.commands` has structured `signatures`, so the output is never like these lines!

> If we're trying to exactly match `help commands` we'll also have to add `command_type`.
>
outside of this PR, right?

**Comment by amtoine** [comment]
@fdncred 
highlighting added in 3305e8fa7056033acb9a0c4a2986a1c5be82c037 with `find $find` :+1: 

> **Warning**
> now, any of the `help` command searches for the `find` pattern in ALL the columns of the selected `$nu.scope.`

**Comment by fdncred** [comment]
I think we need to document the differences between the stdlib help and rust help. I'm pushing to have them be identical in functionality, but others may disagree. In my mind, if they're not identical, then we have a regression in functionality.

Here's what I know is different so far.
1. Speed - We can land with having slight speed differences because we can always optimize that later.
2. `help -f` highlights too many columns in script form.
3. `help commands` has different columns (no command_type) column and the signatures column is different.

**Comment by amtoine** [comment]
@fdncred 
there are also the points above that are not yet addressed

> :bulb: **LEGEND**
> :large_blue_circle: an idea
> :yellow_circle: -> :orange_circle: -> :red_circle: from low to high issue

1. :large_blue_circle: i currently print the exported environment blocks, see `help help modules`
2. :large_blue_circle: i thought of adding a `--raw` option to allow custom data processing by not expanding the help page found, but instead return the raw data
3. :yellow_circle: with the builtin, `help externs open` returns the `help open` even though `open` is not an external => the new `std help externs open` does not
4. :red_circle: in the built-in `help externs`, we have all the registered flags that show up => as they are not available in `$nu`, i do not know how to do that

File: crates/nu-utils/standard_library/std.nu [hunk 1]
```diff
@@ -296,3 +296,721 @@ export def "log debug" [message: string] {
...
+    print ""
+
+    if not ($module.commands? | is-empty) {
+        print-help-header "Exported commands"
+        print -n "    "
+
+        let commands_string = (
+            $module.commands
+            | each {|command|
+                $"($command) (char lparen)(ansi cyan_bold)($module.name) ($command)(ansi reset)(char rparen)"
```
Comment (kubouch) [inline crates/nu-utils/standard_library/std.nu:365]: This one is not correct. I think it's better to drop the cyan version for now because I don't think there is a way to query it correctly from inside Nushell.

The bold cyan shows the actual name of the exported command inside the current scope which depends on how you import it. `use std` will have `std` prefix while `use std *` won't. Later, we might support renaming of modules which would lead to a completely different prefix. We'd probably need to expose the ID of commands inside `$nu` in order to be able to link the module's command with an existing symbol in the current scope.

File: crates/nu-utils/standard_library/std.nu [hunk 2]
```diff
@@ -296,3 +296,721 @@ export def "log debug" [message: string] {
...
+        print-help-header -n "Exported aliases:"
+        print $module.aliases
+        print ""
+    }
+
+    if ($module.env_block? | is-empty) {
+        print $"This module (ansi cyan)does not export(ansi reset) environment."
+    } else {
+        print $"This module (ansi cyan)exports(ansi reset) environment."
+        view source $module.env_block | nu-highlight
```
Comment (kubouch) [inline crates/nu-utils/standard_library/std.nu:384]: This one currently shows as a huge red blob. Prepending `export-env` before the block should fix it. Although I'm thinking if the `export-env` block is too big, it might be quite annoying (see https://github.com/pypa/virtualenv/blob/main/src/virtualenv/activation/nushell/activate.nu, for example), we can just leave it out as well. You can use `view source module-name` to view the source code if you want.

File: crates/nu-utils/standard_library/std.nu (hunk 1 above)
Comment (amtoine) [reply crates/nu-utils/standard_library/std.nu]: should be good in 3c54ca6b3dced82bbb3641c7803e7a4197045fc3

File: crates/nu-utils/standard_library/std.nu (hunk 2 above)
Comment (amtoine) [reply crates/nu-utils/standard_library/std.nu:384]: mm you are rigth
```bash
use crates/nu-utils/standard_library/std.nu 'help modules'
help modules std
```
gives a red blob :thinking:

Comment (amtoine) [reply crates/nu-utils/standard_library/std.nu]: i removed it for now in 1e14b4620b5b3b4fa406c381e3a72e72581ffde5

**Comment by kubouch** [comment]
Oops, sorry, I misclicked.

Looks good to me. It might need some touches but we can improve it as we go. For example, I personally wouldn't show the `export-env` block but that's a detail.
